File: tests/cross_driver/node/rbac_smoke.js
// Cross-driver RBAC smoke — Node (mongo-node-driver).
//
// Same shape as tests/cross_driver/go/rbac/main.go. Provisions a
// `read`-bound user via the root admin connection, then asserts find
// works and insert is rejected with `Unauthorized` when authenticated
// as the new user. The node driver's MongoServerError surfaces the
// errmsg / code from the reply doc — anything other than 13 trips here.

const { MongoClient } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function clientFor(uri, user, pwd, authSource) {
  return new MongoClient(uri, {
    auth: { username: user, password: pwd },
    authSource,
    authMechanism: 'SCRAM-SHA-256',
    serverSelectionTimeoutMS: 5000,
  });
}

async function main() {
  const uri = process.env.MONGODB_URI;
  const adminPwd = process.env.ADMIN_PASSWORD;
  if (!uri || !adminPwd) fail('MONGODB_URI and ADMIN_PASSWORD required');

  const root = clientFor(uri, 'root', adminPwd, 'admin');
  await root.connect();
  try {
    await root
      .db('shop')
      .command({ createUser: 'viewer', pwd: 'vp', roles: [{ role: 'read', db: 'shop' }] });
  } finally {
    await root.close();
  }

  const viewer = clientFor(uri, 'viewer', 'vp', 'shop');
  await viewer.connect();
  try {
    const items = viewer.db('shop').collection('items');
    await items.find({}).toArray();

    let caught = null;
    try {
      await items.insertOne({ x: 1 });
    } catch (err) {
      caught = err;
    }
    if (!caught) fail('insert should have been rejected');
    const message = (caught.errmsg || caught.message || '').toString();
    if (caught.code !== 13 && !message.includes('Unauthorized')) {
      fail(`unexpected error: code=${caught.code} message=${message}`);
    }
    console.log('OK');
  } finally {
    await viewer.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));

File: tests/cross_driver/ruby/rbac_smoke.rb
# Cross-driver RBAC smoke — Ruby (mongo-ruby-driver).
#
# Provisions a ``read``-bound user via the root admin connection,
# then asserts find works and insert is rejected with code 13 /
# Unauthorized when authenticated as the new user.
require "mongo"

def fail_with(msg)
  warn(msg)
  exit(1)
end

uri = ENV["MONGODB_URI"] or fail_with("MONGODB_URI not set")
admin_pwd = ENV["ADMIN_PASSWORD"] or fail_with("ADMIN_PASSWORD not set")
Mongo::Logger.logger.level = Logger::FATAL

root = Mongo::Client.new(
  uri,
  database: "admin",
  user: "root",
  password: admin_pwd,
  auth_source: "admin",
  auth_mech: :scram256,
  server_selection_timeout: 5,
)
begin
  root.use("shop").database.command(
    createUser: "viewer_ruby",
    pwd: "vp",
    roles: [{role: "read", db: "shop"}],
  )
ensure
  root.close
end

viewer = Mongo::Client.new(
  uri,
  database: "shop",
  user: "viewer_ruby",
  password: "vp",
  auth_source: "shop",
  auth_mech: :scram256,
  server_selection_timeout: 5,
)
begin
  # Read should succeed.
  viewer["items"].find({}).to_a

  caught = nil
  begin
    viewer["items"].insert_one({"x" => 1})
  rescue Mongo::Error::OperationFailure => e
    caught = e
  end
  fail_with("insert should have been rejected") unless caught
  unless caught.code == 13 || caught.message.include?("Unauthorized")
    fail_with("unexpected error: code=#{caught.code} message=#{caught.message}")
  end

  puts "OK"
ensure
  viewer.close
end
